libbpfgo.go

```diff
@@ -578,6 +578,116 @@ func (b *BPFMap) GetValue(key unsafe.Pointer) ([]byte, error) {
 	return value, nil
 }
 
+// BPFMapBatchOpts mirrors the C structure bpf_map_batch_opts.
+type BPFMapBatchOpts struct {
+	Sz        uint64
+	ElemFlags uint64
+	Flags     uint64
+}
+
+func bpfMapBatchOptsToC(batchOpts *BPFMapBatchOpts) *C.struct_bpf_map_batch_opts {
+	if batchOpts == nil {
+		return nil
+	}
+	opts := C.struct_bpf_map_batch_opts{}
+	opts.sz = C.ulong(batchOpts.Sz)
+	opts.elem_flags = C.ulonglong(batchOpts.ElemFlags)
+	opts.flags = C.ulonglong(batchOpts.Flags)
+
+	return &opts
+}
+
+// GetValueBatch allows for batch lookups of multiple keys.
+// The first argument is a pointer to an array of keys which will be populated with the keys returned from this operation.
+// It returns the associated values as a slice of slices of bytes.
+func (b *BPFMap) GetValueBatch(keys unsafe.Pointer, startKey, nextKey unsafe.Pointer, count uint32) ([][]byte, error) {
+	var (
+		values    = make([]byte, b.ValueSize()*int(count))
+		valuesPtr = unsafe.Pointer(&values[0])
+		countC    = C.uint(count)
+	)
+
+	opts := &BPFMapBatchOpts{
+		Sz:        uint64(unsafe.Sizeof(BPFMapBatchOpts{})),
+		ElemFlags: C.BPF_ANY,
+		Flags:     C.BPF_ANY,
+	}
+
+	errC := C.bpf_map_lookup_batch(b.fd, startKey, nextKey, keys, valuesPtr, &countC, bpfMapBatchOptsToC(opts))
+	if errC != 0 {
+		return nil, fmt.Errorf("failed to batch lookup values %v in map %s: %d", keys, b.name, errC)
+	}
```

Comment:
errno needs to be checked here, you can use syscall.Errno like elsewhere in libbpfgo.

Reply:
Apologies, I followed the example of the other BPFMap methods (such as `Pin`, `UnPin`, etc...) which don't do any extra checking on the returned error than what I've done. Do you have another specific example or pattern you'd like me to follow or extra checks to perform?

Reply:
I had an answer ready to this but then after carefully examining libbpf code I got confused. I asked on the mailing list how functions which use `libbpf_err` vs `libbpf_err_errno` are supposed to be checked, let's wait for a response from Andrii before proceededing:

https://lore.kernel.org/bpf/CAO658oUqd5=B3zkDhm2jVQxG+vEf=2CE7WimXHqgcH+m0P=k_Q@mail.gmail.com/T/#u

I'm sure there's functions in libbpfgo that aren't correct.

Reply:
@yanivagman You probably have a good idea

Reply:
Ack, sounds good to me!

Reply:
If I understand correctly it looks like Andrii responded saying that the returned value can be used directly without checking errno unless it's an API which returns a pointer, in which case errno must be checked.

Are you reading the response the same?

Reply:
Yes I read it the same, which means the way you're doing it is correct. Though perhaps we can extract the error code to determine what the issue was. I think that's what using the `syscall.Errno` type does, but I will investigate in the morning.